File: src/Blatt_6/time.cpp
#include <iostream>
#include <vector>
#include <chrono>
#include <random>
#include <iomanip>
#include <algorithm>

class Matrix {
private:
    int n;
    std::vector<double> data;
    
public:
    Matrix(int size) : n(size), data(size * size, 0.0) {}
    
    double get(int x, int y) const {
        if (x < 0 || x >= n || y < 0 || y >= n) {
            std::cerr << "Fehler: Index (" << x << ", " << y << ") außerhalb der Grenzen!" << std::endl;
            return 0.0;
        }
        return data[x * n + y];
    }
    
    void set(int x, int y, double value) {
        if (x < 0 || x >= n || y < 0 || y >= n) {
            std::cerr << "Fehler: Index (" << x << ", " << y << ") außerhalb der Grenzen!" << std::endl;
            return;
        }
        data[x * n + y] = value;
    }
    
    int get_dimension() const {
        return n;
    }
    
    void set_identity() {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                set(i, j, (i == j) ? 1.0 : 0.0);
            }
        }
    }
    
    void set_zero() {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                set(i, j, 0.0);
            }
        }
    }
    
    void set_random() {
        static std::random_device rd;
        static std::mt19937 gen(rd());
        static std::uniform_real_distribution<> dis(0.0, 1.0);
        
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                set(i, j, dis(gen));
            }
        }
    }
    
    Matrix multiply(const Matrix& other) const {
        Matrix result(n);
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                double sum_val = 0.0;
                for (int k = 0; k < n; k++) {
                    sum_val += get(i, k) * other.get(k, j);
                }
                result.set(i, j, sum_val);
            }
        }
        return result;
    }
    
    // Optimierte Version mit direktem Datenzugriff
    Matrix multiply_optimized(const Matrix& other) const {
        Matrix result(n);
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                double sum_val = 0.0;
                for (int k = 0; k < n; k++) {
                    sum_val += data[i * n + k] * other.data[k * n + j];
                }
                result.data[i * n + j] = sum_val;
            }
        }
        return result;
    }
    
    // Alternative Schleifenreihenfolge (i-k-j statt i-j-k)
    Matrix multiply_ikj(const Matrix& other) const {
        Matrix result(n);
        result.set_zero();
        for (int i = 0; i < n; i++) {
            for (int k = 0; k < n; k++) {
                for (int j = 0; j < n; j++) {
                    result.data[i * n + j] += data[i * n + k] * other.data[k * n + j];
                }
            }
        }
        return result;
    }
    
    void print_matrix() const {
        std::cout << "Matrix " << n << "x" << n << ":" << std::endl;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                std::cout << std::fixed << std::setprecision(2) << get(i, j) << "\t";
            }
            std::cout << std::endl;
        }
        std::cout << std::endl;
    }
};

void test_matrix_basics() {
    std::cout << "=== C++ Matrix-Klasse Test ===" << std::endl;
    
    Matrix m(3);
    std::cout << "Dimension: " << m.get_dimension() << std::endl;
    
    // Werte setzen
    m.set(0, 0, 1.0); m.set(0, 1, 2.0); m.set(0, 2, 3.0);
    m.set(1, 0, 4.0); m.set(1, 1, 5.0); m.set(1, 2, 6.0);
    m.set(2, 0, 7.0); m.set(2, 1, 8.0); m.set(2, 2, 9.0);
    
    m.print_matrix();
    
    std::cout << "m[1][1] = " << m.get(1, 1) << std::endl;
    std::cout << "m[0][2] = " << m.get(0, 2) << std::endl;
    std::cout << std::endl;
}

void test_matrix_multiplication() {
    std::cout << "=== Matrix-Multiplikation Test ===" << std::endl;
    
    Matrix a(2);
    Matrix b(2);
    
    // Matrix A: [[1, 2], [3, 4]]
    a.set(0, 0, 1); a.set(0, 1, 2);
    a.set(1, 0, 3); a.set(1, 1, 4);
    
    // Matrix B: [[2, 0], [1, 2]]
    b.set(0, 0, 2); b.set(0, 1, 0);
    b.set(1, 0, 1); b.set(1, 1, 2);
    
    std::cout << "Matrix A:" << std::endl;
    a.print_matrix();
    std::cout << "Matrix B:" << std::endl;
    b.print_matrix();
    
    Matrix result = a.multiply(b);
    std::cout << "A * B =" << std::endl;
    result.print_matrix();
}

void performance_test() {
    std::cout << "=== C++ Performance-Test ===" << std::endl;
    
    std::vector<int> sizes = {10, 50, 100, 200, 500};
    
    for (int size : sizes) {
        Matrix a(size);
        Matrix b(size);
        a.set_identity();
        b.set_identity();
        
        auto start = std::chrono::high_resolution_clock::now();
        Matrix result = a.multiply(b);
        auto end = std::chrono::high_resolution_clock::now();
        
        auto duration = std::chrono::duration_cast<std::chrono::microseconds>(end - start);
        double ms = duration.count() / 1000.0;
        
        std::cout << "Size " << size << "x" << size << ": " 
                  << std::fixed << std::setprecision(2) << ms << " ms" << std::endl;
    }
}

void performance_test_detailed() {
    std::cout << "=== Detaillierte Performance-Messung ===" << std::endl;
    
    std::vector<int> sizes = {10, 50, 100, 200};
    const int runs = 3;
    
    for (int size : sizes) {
        std::vector<double> times;
        
        for (int run = 0; run < runs; run++) {
            Matrix a(size);
            Matrix b(size);
            a.set_random();
            b.set_random();
            
            auto start = std::chrono::high_resolution_clock::now();
            Matrix result = a.multiply(b);
            auto end = std::chrono::high_resolution_clock::now();
            
            auto duration = std::chrono::duration_cast<std::chrono::microseconds>(end - start);
            times.push_back(duration.count() / 1000.0);
        }
        
        double avg_time = 0;
        for (double t : times) avg_time += t;
        avg_time /= times.size();
        
        double min_time = *std::min_element(times.begin(), times.end());
        double max_time = *std::max_element(times.begin(), times.end());
        
        std::cout << "Size " << size << "x" << size << ": "
                  << "avg=" << std::fixed << std::setprecision(2) << avg_time << "ms, "
                  << "min=" << min_time << "ms, "
                  << "max=" << max_time << "ms" << std::endl;
    }
}

void compare_algorithms() {
    std::cout << "=== Algorithmen-Vergleich ===" << std::endl;
    
    const int size = 100;
    Matrix a(size);
    Matrix b(size);
    a.set_random();
    b.set_random();
    
    // Standard i-j-k
    auto start = std::chrono::high_resolution_clock::now();
    Matrix result1 = a.multiply(b);
    auto end = std::chrono::high_resolution_clock::now();
    double time_ijk = std::chrono::duration_cast<std::chrono::microseconds>(end - start).count() / 1000.0;
    
    // Optimiert mit direktem Datenzugriff
    start = std::chrono::high_resolution_clock::now();
    Matrix result2 = a.multiply_optimized(b);
    end = std::chrono::high_resolution_clock::now();
    double time_opt = std::chrono::duration_cast<std::chrono::microseconds>(end - start).count() / 1000.0;
    
    // Alternative i-k-j
    start = std::chrono::high_resolution_clock::now();
    Matrix result3 = a.multiply_ikj(b);
    end = std::chrono::high_resolution_clock::now();
    double time_ikj = std::chrono::duration_cast<std::chrono::microseconds>(end - start).count() / 1000.0;
    
    std::cout << "Standard (i-j-k): " << std::fixed << std::setprecision(2) << time_ijk << " ms" << std::endl;
    std::cout << "Optimiert (direkter Zugriff): " << time_opt << " ms" << std::endl;
    std::cout << "Alternative (i-k-j): " << time_ikj << " ms" << std::endl;
    std::cout << "Speedup Standard vs Optimiert: " << time_ijk/time_opt << "x" << std::endl;
    std::cout << "Speedup Standard vs i-k-j: " << time_ijk/time_ikj << "x" << std::endl;
}

void benchmark_comparison() {
    std::cout << "\n=== Python vs C++ Benchmark ===" << std::endl;
    std::cout << "Größe\tC++ Zeit\tPython Zeit (geschätzt)\tSpeedup" << std::endl;
    std::cout << "------------------------------------------------" << std::endl;
    
    std::vector<int> sizes = {10, 50, 100, 200, 500};
    
    for (int size : sizes) {
        Matrix a(size);
        Matrix b(size);
        a.set_random();
        b.set_random();
        
        // Beste C++ Version verwenden (i-k-j mit direktem Zugriff)
        auto start = std::chrono::high_resolution_clock::now();
        Matrix result = a.multiply_ikj(b);
        auto end = std::chrono::high_resolution_clock::now();
        
        double cpp_time = std::chrono::duration_cast<std::chrono::microseconds>(end - start).count() / 1000.0;
        
        // Geschätzte Python-Zeit basierend auf typischen Faktoren
        double estimated_python_time = cpp_time * 50; // Python ist typischerweise 50-100x langsamer
        
        std::cout << size << "x" << size << "\t"
                  << std::fixed << std::setprecision(2) << cpp_time << " ms\t"
                  << estimated_python_time << " ms\t\t"
                  << "~" << (int)(estimated_python_time / cpp_time) << "x" << std::endl;
    }
}

int main() {
    test_matrix_basics();
    test_matrix_multiplication();
    performance_test();
    performance_test_detailed();
    compare_algorithms();
    benchmark_comparison();
    
    return 0;
}

File: src/Blatt_6/time.py
import time
import random

class Matrix:
    def __init__(self, size):
        self.n = size
        self.data = [0.0] * (size * size)  # 1D Liste wie in C++
    
    def get(self, x, y):
        if x < 0 or x >= self.n or y < 0 or y >= self.n:
            print(f"Fehler: Index ({x}, {y}) außerhalb der Grenzen!")
            return 0.0
        return self.data[x * self.n + y]
    
    def set(self, x, y, value):
        if x < 0 or x >= self.n or y < 0 or y >= self.n:
            print(f"Fehler: Index ({x}, {y}) außerhalb der Grenzen!")
            return
        self.data[x * self.n + y] = value
    
    def get_dimension(self):
        return self.n
    
    def set_identity(self):
        """Setzt die Matrix als Einheitsmatrix"""
        for i in range(self.n):
            for j in range(self.n):
                self.set(i, j, 1.0 if i == j else 0.0)
    
    def set_zero(self):
        """Setzt alle Werte auf 0"""
        for i in range(self.n):
            for j in range(self.n):
                self.set(i, j, 0.0)
    
    def set_random(self):
        """Füllt die Matrix mit Zufallswerten zwischen 0 und 1"""
        for i in range(self.n):
            for j in range(self.n):
                self.set(i, j, random.random())
    
    def multiply(self, other):
        """Matrix-Multiplikation - gleiche Logik wie C++"""
        result = Matrix(self.n)
        for i in range(self.n):
            for j in range(self.n):
                sum_val = 0.0
                for k in range(self.n):
                    sum_val += self.get(i, k) * other.get(k, j)
                result.set(i, j, sum_val)
        return result
    
    def print_matrix(self):
        """Matrix ausgeben für Debugging"""
        print(f"Matrix {self.n}x{self.n}:")
        for i in range(self.n):
            for j in range(self.n):
                print(f"{self.get(i, j):.2f}", end="\t")
            print()
        print()

def test_matrix_basics():
    """Grundlegende Tests der Matrix-Klasse"""
    print("=== Python Matrix-Klasse Test ===")
    
    # Matrix erstellen
    m = Matrix(3)
    print(f"Dimension: {m.get_dimension()}")
    
    # Werte setzen
    m.set(0, 0, 1.0)
    m.set(0, 1, 2.0)
    m.set(0, 2, 3.0)
    m.set(1, 0, 4.0)
    m.set(1, 1, 5.0)
    m.set(1, 2, 6.0)
    m.set(2, 0, 7.0)
    m.set(2, 1, 8.0)
    m.set(2, 2, 9.0)
    
    # Matrix ausgeben
    m.print_matrix()
    
    # Einzelne Werte abfragen
    print(f"m[1][1] = {m.get(1, 1)}")
    print(f"m[0][2] = {m.get(0, 2)}")
    print()

def test_matrix_multiplication():
    """Test der Matrix-Multiplikation"""
    print("=== Matrix-Multiplikation Test ===")
    
    # Zwei 2x2 Matrizen erstellen
    a = Matrix(2)
    b = Matrix(2)
    
    # Matrix A: [[1, 2], [3, 4]]
    a.set(0, 0, 1); a.set(0, 1, 2)
    a.set(1, 0, 3); a.set(1, 1, 4)
    
    # Matrix B: [[2, 0], [1, 2]]
    b.set(0, 0, 2); b.set(0, 1, 0)
    b.set(1, 0, 1); b.set(1, 1, 2)
    
    print("Matrix A:")
    a.print_matrix()
    print("Matrix B:")
    b.print_matrix()
    
    # Multiplikation
    result = a.multiply(b)
    print("A * B =")
    result.print_matrix()
    
    # Ergebnis sollte sein: [[4, 4], [10, 8]]

def performance_test():
    """Performance-Test für verschiedene Matrixgrößen"""
    print("=== Python Performance-Test ===")
    
    sizes = [10, 50, 100, 200, 500]
    
    for size in sizes:
        # Zwei Einheitsmatrizen erstellen
        a = Matrix(size)
        b = Matrix(size)
        a.set_identity()
        b.set_identity()
        
        # Zeitmessung
        start_time = time.time()
        result = a.multiply(b)
        end_time = time.time()
        
        duration_ms = (end_time - start_time) * 1000
        print(f"Size {size}x{size}: {duration_ms:.2f} ms")
        
        # Bei großen Matrizen wird es sehr langsam
        if size >= 500:
            print("  (Warnung: Sehr langsam!)")

def performance_test_detailed():
    """Detaillierte Performance-Messung mit mehreren Durchläufen"""
    print("=== Detaillierte Performance-Messung ===")
    
    sizes = [10, 50, 100, 200]
    runs = 3  # Mehrere Durchläufe für genauere Messung
    
    for size in sizes:
        times = []
        
        for run in range(runs):
            # Zufällige Matrizen für realistischeren Test
            a = Matrix(size)
            b = Matrix(size)
            a.set_random()
            b.set_random()
            
            start_time = time.time()
            result = a.multiply(b)
            end_time = time.time()
            
            times.append((end_time - start_time) * 1000)
        
        avg_time = sum(times) / len(times)
        min_time = min(times)
        max_time = max(times)
        
        print(f"Size {size}x{size}: avg={avg_time:.2f}ms, min={min_time:.2f}ms, max={max_time:.2f}ms")

def compare_algorithms():
    """Vergleich verschiedener Matrix-Multiplikationsalgorithmen"""
    print("=== Algorithmen-Vergleich ===")
    
    class MatrixOptimized(Matrix):
        def multiply_ikj(self, other):
            """Alternative Schleifenreihenfolge (i-k-j statt i-j-k)"""
            result = Matrix(self.n)
            for i in range(self.n):
                for k in range(self.n):
                    for j in range(self.n):
                        current = result.get(i, j)
                        result.set(i, j, current + self.get(i, k) * other.get(k, j))
            return result
    
    size = 100
    a = MatrixOptimized(size)
    b = MatrixOptimized(size)
    a.set_random()
    b.set_random()
    
    # Standard i-j-k
    start = time.time()
    result1 = a.multiply(b)
    time_ijk = (time.time() - start) * 1000
    
    # Alternative i-k-j
    start = time.time()
    result2 = a.multiply_ikj(b)
    time_ikj = (time.time() - start) * 1000
    
    print(f"Standard (i-j-k): {time_ijk:.2f} ms")
    print(f"Alternative (i-k-j): {time_ikj:.2f} ms")
    print(f"Speedup: {time_ijk/time_ikj:.2f}x")

if __name__ == "__main__":
    test_matrix_basics()
    test_matrix_multiplication()
    performance_test()
    performance_test_detailed()
    compare_algorithms()